### PR TITLE
[Android NNAPI EP] Enable test for BatchNormalization, enable dev_mode for Android, fix some issues in concat

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Enable test for BatchNormalization, enable dev_mode for Android, fix some issues in concat

**Motivation and Context**
- Disabled the deprecated API warning for ONNX, enable the warning as error for Android compile settings
- Enabled the test for BatchNorm
- Fixed some issue for dynamic input support for concat
- Fixed issue where NNAPI will fail if the input has no shape (not a scalar)

### CODE REVIEW

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc [hunk 1]
```diff
@@ -1195,14 +1221,17 @@ Status BatchNormalizationOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_bu
...
   bool input_is_nhwc = model_builder.IsOperandNHWC(input);
   bool output_is_nhwc = input_is_nhwc;
-  if (input_is_nhwc)
-    tensor_a_dimen = {size};
-  else                              // input is nchw
-    tensor_a_dimen = {size, 1, 1};  // {C, H, W}
+
+  if (!input_is_nhwc) {
+    // input shape is {N, C, [H, W]}, so tensor_a/b's shape is {C, [H, W]}
+    const auto input_rank = shaper[input].size();
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:1231]: Not quite clear 1 is pushed for the H and W dims if input is nchw so a comment explaining why would be helpful.

**Comment by skottmckay** [comment]
Was the Concat issue caught by a failing test or manually? If manually can we add a unit test that would have triggered the problem?

**Comment by gwang-msft** [comment]
The issue was caught by a real world model with one of the inputs which has a dynamic shape, will need to figure out how to do it in a unit test 


---
In reply to: [669677470](https://github.com/microsoft/onnxruntime/pull/4715#issuecomment-669677470) [](ancestors = 669677470)

**Comment by gwang-msft** [comment]
Added a new test case for Concat

---
In reply to: [669685804](https://github.com/microsoft/onnxruntime/pull/4715#issuecomment-669685804) [](ancestors = 669685804,669677470)

File: onnxruntime/test/providers/cpu/tensor/concat_op_test.cc [hunk 2]
```diff
@@ -111,6 +111,24 @@ TEST(ConcatOpTest, Concat2D_3) {
             kNnapiExecutionProvider});   // NNAPI: concat does not support 0 size input
 }
 
+TEST(ConcatOpTest, Concat2D_4) {
+  OpTester test("Concat");
+  test.AddAttribute("axis", int64_t{1});
+
+  std::vector<int64_t> dims{4, 1};
+  std::vector<std::string> dim_params{"batch", "seq"};
+  test.AddInput<float>("input1", dims, {11.0f, 21.0f, 31.0f, 41.0f});
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/tensor/concat_op_test.cc:120]: nit: would be helpful to put a comment explaining why we do something so obtusely different for one input. that way the next developer to look at the code knows it's intentional and why it's being done, as it's unlikely they're aware of the specific processing in the NNAPI EP being tested.

**Pull request merged**